<a href="https://colab.research.google.com/github/ChikaLong/Python/blob/main/pandas_editor.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# pandas exercises


1. データの読み込み方法について学習します。(2問)
2. データの状態を確認する方法について学習します。（11問）
3. データの条件を指定する方法について学習します。（11問）
4. データの集計方法について学習します。（7問）
5. データの整形方法について学習します。（13問）
6. 文字列の整形について学習します。（6問）

### 0. 事前準備

#### 0.1 「pandasをpd、numpyをnpと読み込んでください。」

In [49]:
import pandas as pd
import numpy as np

#### 0.2 「次のコードを実行してください。※data変数に格納してください。」

In [50]:
data = pd.DataFrame({'name': ['A', 'B', 'C', 'D'], '数学': [80, 15, 90, 50], '英語': [80, 70, 50, 65],
                    '国語': [ 90, 60, 60,  60]})

#### 0.3 「次のコードを実行してください。※e_data変数に格納してください。」

In [51]:
e_data = pd.DataFrame({'name':'E','国語':[30],'数学':[30],'英語':[30]})

#### 0.4 「次のコードを実行してください。※data_2変数に格納してください。」

In [52]:
data_2 = pd.DataFrame({'name': ['A', 'B', 'C', 'D','D'], '国語-数学': ["30:20", np.NAN, "90:50", "50:30", "50:30"], '英語': ["80.2","60.5","34,2",0,0]})

#### 0.5 「次のコードを実行してください。※data_3変数に格納してください。」

In [53]:
data_3 = pd.DataFrame({'name': ['A', 'B', 'C', 'D'],
                     '性別': ['男性', '男性', '女性', '男性'],
                     '経歴': ['学部卒', '修士卒', '学部卒', '博士卒'],
                     '特技': ['ボーリング', '野球', 'サッカー', '野球をやっていた']})

### 1. データを読み込む方法について学習します。（2問）

#### 1.1 「data変数を現在のディレクトリ配下に「a.csv」という名前で保存してください。」

In [54]:
data.to_csv('a.csv', index=False)

A. **data.to_csv('./a.csv', index=False)**

#### 1.2「「a.csv」というファイルを読み込み、dataという名前の変数に格納してください。」

In [55]:
data = pd.read_csv('./a.csv')
data.tail()

name  数学  英語  国語
0    A  80  80  90
1    B  15  70  60
2    C  90  50  60
3    D  50  65  60

A. **data = pd.read_csv('./a.csv', index_col = False)**


### 2.データの状態を確認する方法について学習します。（11問）

#### 2.1 「data変数のデータ形式を表示してください。」

In [56]:
type(data[:])

pandas.core.frame.DataFrame

#### 2.2 「pandasでSeries型データを作成して、データ形式を確認してください。※Seriesの中身は何でも良いです。」

In [57]:
sr = pd.Series([100,200,300])
type(sr)

pandas.core.series.Series

#### 2.3 「pandasでDataFrame型データを作成して、データ形式を確認してください。※DataFrameの中身は何でも良いです。」

In [58]:
df = pd.DataFrame([100,200,300])
type(df)

pandas.core.frame.DataFrame

#### 2.4 「data変数の行と列を表示してください。」

In [59]:
data.shape

(4, 4)

#### 2.5 「data変数の上から2行表示してください。」

In [60]:
data.head(2)

name  数学  英語  国語
0    A  80  80  90
1    B  15  70  60

#### 2.6 「data変数の下から3行表示してください。」

In [61]:
data.tail(3)

name  数学  英語  国語
1    B  15  70  60
2    C  90  50  60
3    D  50  65  60

#### 2.7 「data変数のカラムを表示してください。」

In [62]:
data.columns

Index(['name', '数学', '英語', '国語'], dtype='object')

#### 2.8 「data変数のインデックスを表示してください。」

In [63]:
data.index

RangeIndex(start=0, stop=4, step=1)

#### 2.9 「data変数の各カラムのdata型を表示してください。」

In [64]:
data.dtypes

name    object
数学       int64
英語       int64
国語       int64
dtype: object

#### 2.10 「data変数の各カラムのユニーク数を表示してください。」

In [65]:
data.nunique()

name    4
数学      4
英語      4
国語      2
dtype: int64

#### 2.11 「data変数の各インデックスのユニーク数を表示してください。」

In [66]:
data.nunique(axis=1)

0    3
1    4
2    4
3    4
dtype: int64

### 3.データの条件を指定する方法について学習します。（11問）

#### 3.1 「data変数の上から4行目のデータを表示してください。」

In [67]:
data.loc[3,:]

name     D
数学      50
英語      65
国語      60
Name: 3, dtype: object

#### 3.2 「data変数の中で国語が70点以上の人を表示してください。」

In [68]:
data.query('国語 >= 70')

name  数学  英語  国語
0    A  80  80  90

A. **data[data['国語'] >= 70]**

#### 3.3 「data変数の中で数学が30点未満の人を表示してください。」

In [69]:
data.query('数学 < 30')

name  数学  英語  国語
1    B  15  70  60

A. **data[data['数学'] < 30]**

#### 3.4 「data変数の中で国語が60点ピッタリの人が何人いるか表示してください。」

In [70]:
(data['国語'] == 60).sum()

3

A. **len(data[data['国語'] == 60])**

#### 3.5 「data変数の中で数学が50点以上の人が何人いるか表示してください。」

In [71]:
(data['数学'] >= 50).sum()

3

A. **len(data[data['数学'] >= 50])**

#### 3.6「data変数の中で国語の点数が1番高い人を表示してください。」

In [72]:
data[data['国語'] == data['国語'].max()]['name']

0    A
Name: name, dtype: object

#### 3.7 「data変数の中で全ての人の国語の点数を表示してください。」

In [73]:
data['国語']

0    90
1    60
2    60
3    60
Name: 国語, dtype: int64

A. **data.loc[:,'国語']**

#### 3.8 「data変数の中で全ての人の国語と数学の点数を表示してください。」

In [74]:
data[['国語', '数学']][:]

国語  数学
0  90  80
1  60  15
2  60  90
3  60  50

A. **data[['国語','数学']]**

#### 3.9 「data変数のカラムの情報を確認してください。」

In [75]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 4 entries, 0 to 3
Data columns (total 4 columns):
 #   Column  Non-Null Count  Dtype 
---  ------  --------------  ----- 
 0   name    4 non-null      object
 1   数学      4 non-null      int64 
 2   英語      4 non-null      int64 
 3   国語      4 non-null      int64 
dtypes: int64(3), object(1)
memory usage: 256.0+ bytes


#### 3.10 「data変数の中で2行目と3行目のみ表示してください。(indexは0番目が1行目であることに注意)」

In [76]:
data[1:3]

name  数学  英語  国語
1    B  15  70  60
2    C  90  50  60

#### 3.11 「data変数の中で2行目と3行目且つ0列目と1列目を表示してください。(indexは0番目が1行目であることに注意)」

In [77]:
data.iloc[1:3,0:2]

name  数学
1    B  15
2    C  90

A. **data.iloc[1:3,[0,1]]**

### 4. データの集計方法について学習します。（7問）

#### 4.1 「dataの下にe_dataを連結させて下さい。dataの中身を確認してください。」

In [78]:
data = pd.concat([data,e_data],axis=0,sort=True)
data

name  国語  数学  英語
0    A  90  80  80
1    B  60  15  70
2    C  60  90  50
3    D  60  50  65
0    E  30  30  30

#### 4.2 「dataのindexを振り直してください。」

In [79]:
data = data.reset_index(drop=True)
data

name  国語  数学  英語
0    A  90  80  80
1    B  60  15  70
2    C  60  90  50
3    D  60  50  65
4    E  30  30  30

A. **data = data.reset_index()<br>
&emsp;data.drop(['index'],axis=1,inplace=True)<br>
&emsp;data**

#### 4.3 「dataの各カラムの合計値を表示して下さい。」

In [80]:
data.sum()

name    ABCDE
国語        300
数学        265
英語        295
dtype: object

#### 4.4 「dataの各行の合計値を表示して下さい。」

In [81]:
data.sum(axis = 1)

0    250
1    145
2    200
3    175
4     90
dtype: int64

#### 4.5 「dataの各行の平均値を表示して下さい。」

In [82]:
data.mean(axis = 1)

0    83.333333
1    48.333333
2    66.666667
3    58.333333
4    30.000000
dtype: float64

#### 4.6 「data変数のカラムの統計量を算出してください。（小数点第1桁まで）」

In [83]:
data.describe().round()

国語    数学    英語
count   5.0   5.0   5.0
mean   60.0  53.0  59.0
std    21.0  32.0  19.0
min    30.0  15.0  30.0
25%    60.0  30.0  50.0
50%    60.0  50.0  65.0
75%    60.0  80.0  70.0
max    90.0  90.0  80.0

A. **data.describe().round(1)**

#### 4.7 「data変数の中の各カラムの相関係数を算出してください。」

In [84]:
data.corr()

国語        数学        英語
国語  1.000000  0.553509  0.906845
数学  0.553509  1.000000  0.186724
英語  0.906845  0.186724  1.000000

### 5. データの整形方法について学習します。（13問）

#### 5.1 「data変数の数学カラムのユニーク値を表示してください。」

In [85]:
data['数学'].unique()

array([80, 15, 90, 50, 30])

#### 5.2 「data変数の中で数学カラムの中に重複がないことを確認してください。（等式を作成してください。）」

In [86]:
len(data) == len(data['数学'].unique())

True

A. **len(data) == len(data['数学'].unique())**

#### 5.3 「data変数の中で国語カラムの中に重複があることを確認してください。（等式を作成してください。）」

In [87]:
len(data) == len(data['国語'].unique())

False

A. **len(data) == len(data['国語'].unique())**

#### 5.4 「data変数の中で国語の点数が60点の人が3人います。重複レコードを削除してください。」

In [88]:
data[~data['国語'].duplicated(keep='first')]

name  国語  数学  英語
0    A  90  80  80
1    B  60  15  70
4    E  30  30  30

A. **data[~data['国語'].duplicated()]**

#### 5.5 「data_2を表示して下さい。」

In [89]:
data_2.tail()

name  国語-数学    英語
0    A  30:20  80.2
1    B    NaN  60.5
2    C  90:50  34,2
3    D  50:30     0
4    D  50:30     0

#### 5.6 「名前の行に重複が見られます。その行を削除してください。」

In [90]:
data_2 = data_2.drop(data_2.index[4])
data_2

name  国語-数学    英語
0    A  30:20  80.2
1    B    NaN  60.5
2    C  90:50  34,2
3    D  50:30     0

A. **data_2.drop_duplicates('name', inplace=True)<br>
&emsp;data_2**

#### 5.7 「国語-数学の列を国語と数学に分けた列を作成してください。」

A. **data_2['split'] = data_2['国語-数学'].str.split(':')<br>
&emsp;data_2['国語'] = data_2['split'].str.get(0)<br>
&emsp;data_2['数学'] = data_2['split'].str.get(1)<br>
&emsp;data_2**

In [91]:
data_2['split'] = data_2['国語-数学'].str.split(':')
data_2['国語'] = data_2['split'].str.get(0)
data_2['数学'] = data_2['split'].str.get(1)
data_2

name  国語-数学    英語     split   国語   数学
0    A  30:20  80.2  [30, 20]   30   20
1    B    NaN  60.5       NaN  NaN  NaN
2    C  90:50  34,2  [90, 50]   90   50
3    D  50:30     0  [50, 30]   50   30

#### 5.8 「name、英語、国語、数学以外の列を削除してください。」

A. **data_2 = data_2.drop(['国語-数学','split'],axis=1)<br>
&emsp;data_2**

In [92]:
data_2 = data_2.drop(['国語-数学','split'],axis=1)
data_2

name    英語   国語   数学
0    A  80.2   30   20
1    B  60.5  NaN  NaN
2    C  34,2   90   50
3    D     0   50   30

#### 5.9 「Cの英語の点数にカンマが入っています。小数点になおしてください。」

In [93]:
data_2['英語'] = data_2['英語'].str.replace(',' ,  '.').astype(float)
data_2

name    英語   国語   数学
0    A  80.2   30   20
1    B  60.5  NaN  NaN
2    C  34.2   90   50
3    D   NaN   50   30

A. **data_2.loc[2,'英語'] = 34.2<br>
&emsp;data_2**

#### 5.10 「data_2の各列の型(クラス、数字や文字列などの情報)を表示して下さい。」

A. **data_2.dtypes**

In [94]:
data_2.dtypes

name     object
英語      float64
国語       object
数学       object
dtype: object

#### 5.11 「英語、国語、数学を数値型に変更し、データ型が変換されたことを確認してください。」

A. **data_2[['英語','国語','数学']] = data_2[['英語','国語','数学']].astype(float)<br>
&emsp;data_2.dtypes**

In [95]:
data_2[['英語','国語','数学']] = data_2[['英語','国語','数学']].astype(float)
data_2.dtypes

name     object
英語      float64
国語      float64
数学      float64
dtype: object

#### 5.12 「欠損値がある行を表示して下さい。」

A. **data_2.loc[data_2.isnull().any( axis = 1), :]**

In [96]:
data_2.loc[data_2.isnull().any( axis = 1), :]

name    英語    国語    数学
1    B  60.5   NaN   NaN
3    D   NaN  50.0  30.0

#### 5.13 「欠損をもつ行を取り除いて下さい。」

A. **data_2.dropna(axis=0, how = 'any', inplace=True)<br>
&emsp;data_2**

In [97]:
data_2.dropna(axis=0, how = 'any', inplace=True)
data_2

name    英語    国語    数学
0    A  80.2  30.0  20.0
2    C  34.2  90.0  50.0

### 6.文字列の処理（6問）

#### 6.1 「data_3を確認して下さい。」

In [98]:
data_3.tail()

name  性別   経歴        特技
0    A  男性  学部卒     ボーリング
1    B  男性  修士卒        野球
2    C  女性  学部卒      サッカー
3    D  男性  博士卒  野球をやっていた

#### 6.2 「経歴が、'学部卒'で終わる行を表示して下さい。」

In [99]:
data_3[data_3['経歴'].str.endswith('学部卒')]

name  性別   経歴     特技
0    A  男性  学部卒  ボーリング
2    C  女性  学部卒   サッカー

#### 6.3 「性別が、'女'から始まる行を表示して下さい。」

In [100]:
data_3[data_3['性別'].str.startswith('女')]

name  性別   経歴    特技
2    C  女性  学部卒  サッカー

#### 6.4 「特技が'野球'を含む行を表示して下さい。」

In [101]:
data_3[data_3['特技'].str.contains('野球')]

name  性別   経歴        特技
1    B  男性  修士卒        野球
3    D  男性  博士卒  野球をやっていた

#### 6.5 「経歴の'卒'を削除して下さい。」

In [102]:
data_3['経歴'] = data_3['経歴'].str[0:2]
data_3

name  性別  経歴        特技
0    A  男性  学部     ボーリング
1    B  男性  修士        野球
2    C  女性  学部      サッカー
3    D  男性  博士  野球をやっていた

#### 6.6 「特技の'野球をやっていた'の「をやっていた」を削除して下さい。」

In [103]:
data_3['特技'][3] = data_3['特技'][3].strip('をやっていた')
data_3

name  性別  経歴     特技
0    A  男性  学部  ボーリング
1    B  男性  修士     野球
2    C  女性  学部   サッカー
3    D  男性  博士     野球